Sources/SymbolKit/SymbolGraph/Relationship/Relationship.swift

```diff
@@ -116,20 +199,139 @@ extension SymbolGraph.Relationship: Hashable, Equatable {
         hasher.combine(target)
         hasher.combine(kind.rawValue)
         hasher.combine(targetFallback)
-        hasher.combine(mixins[SymbolGraph.Relationship.Swift.GenericConstraints.mixinKey] as? Swift.GenericConstraints)
-        hasher.combine(mixins[SymbolGraph.Relationship.SourceOrigin.mixinKey] as? SourceOrigin)
+        
+        for (key, mixin) in mixins {
+            hasher.combine(key)
+            hasher.combine(mixin.maybeHashable)
+        }
     }
 
     /// A custom equality implmentation for a relationship.
     /// > Important: If there are new relationship mixins they need to be added to the equality function.
     public static func == (lhs: SymbolGraph.Relationship, rhs: SymbolGraph.Relationship) -> Bool {
-        return lhs.source == rhs.source
+        guard lhs.source == rhs.source
             && lhs.target == rhs.target
             && lhs.kind == rhs.kind
             && lhs.targetFallback == rhs.targetFallback
-            && lhs.mixins[SymbolGraph.Relationship.Swift.GenericConstraints.mixinKey] as? Swift.GenericConstraints
-                == rhs.mixins[SymbolGraph.Relationship.Swift.GenericConstraints.mixinKey] as? Swift.GenericConstraints
-            && lhs.mixins[SymbolGraph.Relationship.SourceOrigin.mixinKey] as? SourceOrigin
-                == rhs.mixins[SymbolGraph.Relationship.SourceOrigin.mixinKey] as? SourceOrigin
+            && lhs.mixins.count == rhs.mixins.count else {
+            return false
+        }
+        
+        for (key, lhs) in lhs.mixins {
+            if lhs.maybeEquatable != rhs.mixins[key]?.maybeEquatable {
+                return false
+            }
+        }
+        
+        return true
+    }
+}
+
+private extension Mixin {
+    // A type-erased wrapper around this `Mixin`, which conforms to
+    // `Equatable`.
+    //
+    // If this `Mixin` conforms to `Equatable`, the
+    // `maybeEquatable` uses the `Mixin`'s equality function.
+    // However, if this `Mixin` does not conform to `Equatable`,
+    // the `maybeEquatable` merely checks that the compared elements'
+    // types match.
+    var maybeEquatable: MaybeEquatable {
```

Comment:
`Relationship` is Hashable so that Swift-DocC can build sets of them when resolving external symbols. Equatable is there because Hashable requires it. As far as i can tell, `Symbol` does not implement either of these protocols (though now it could, given that there's a mechanism for `Mixin` to be hashed/equated).

Reply:
> I'm leaning towards SymbolKit checking equality of Equatable mixins defined in SymbolKit, and to document that clients should do equality checking for their own mixins if required.

I disagree. I think this would make handling Symbols/Relationships very annoying for clients with custom mixins. E.g. if you wanted to use the types as a key in a Dictionary, you'd always have to wrap them in a type that implements the additional equality constraints.

What would be a situation where you think my approach would be problematic @franklinsch? Or is this more about code cleanliness/performance concerns?

Reply:
The added flexibility is nice, my concern is more about SymbolKit having an internal expectation that mixins are Equatable when in fact that's not enforced by the type system. Should we just make `Mixin` Equatable and Hashable, then?

Reply:
The problem with making `Mixin` Equatable and Hashable is that it restricts us to Swift 5.7 - before then there's no way to use Mixin generically since the Equatable definition needs to reference the type in question.

Reply:
> my concern is more about SymbolKit having an internal expectation that mixins are Equatable when in fact that's not enforced by the type system

SymbolKit can rightfully make the assumption that everything important for equality will be considered in `Symbol`/`Relationship`'s `Equatable` conformance. SymbolKit knows for itself all predefined Mixins are `Equatable` and the rest is up to those defining custom Mixins.

> Should we just make Mixin Equatable and Hashable, then?

If we were to do that, `Mixin` would get "self or associated type requirements", which (pre Swift 5.7) would break A LOT of source code. I'd only do that if we also bump SymbolKit's minimum language version to 5.7.

Reply:
I'm not opposed to bumping SymbolKit to 5.7 on the `main` branch, which tracks the minor/major release after 5.7. The SymbolKit API is still evolving, and clients that need to use <5.7 can depend on SymbolKit's `release/5.7` branch instead. Also, because SymbolKit is (mostly) a model library that evolves alongside the Swift compiler, it's not unreasonable to align SymbolKit's compiler version requirements with the toolchain version, even though of course, backwards compatibility wherever we can would be better.

Reply:
I guess that would also mean bumping the language version for all of apple/swift-docc to 5.7 then? Don't get me wrong, I'm absolutely not opposed to adopting Swift 5.7 - after all, it would be a huge quality of life improvement for all maintainers. I'm just not sure if getting rid of this (admittedly rather complex and - consequently - very ugly) type-erasure code justifies such a huge breakage.